lib/excon/socket.rb

```diff
@@ -304,17 +311,30 @@ def write_block(data)
     end
 
     def select_with_timeout(socket, type)
+      timeout_kind = type
+      timeout = @data[OPERATION_TO_TIMEOUT[type]]
+
+      if @data.include?(:deadline)
+        remaining = check_deadline!
+
+        if remaining < timeout
+          timeout_kind = :request
+          timeout = remaining
+        end
+      end
+
       select = case type
       when :connect_read
-        IO.select([socket], nil, nil, @data[:connect_timeout])
+        IO.select([socket], nil, nil, timeout)
```

Comment:
Maybe we could simplify this, and the lookup above. An idea I had would be to have a local variable here, something like `request_timeout` which would either be nil or a float if there is time remaining (much as you have above). The difference would be that here in the `IO.select` we could have something like `request_timeout || @data[:connect_timeout]`. That feels like it should be mostly the same, but maybe a little easier to read and shorter. What do you think?

Reply:
I used the `timeout` variable in order to keep some sort of symmetry between the timeout and the kind of timeout. Since we need to update the kind used in the exception messsage if we use the request timeout.

At first I had a conditional in each branch to choose the timeout and kind to use, but that seemed too repetitive. I also tried returning the kind and timeout from the check deadline method but I didn't like the multi-return pattern. Something like `timeout, kind = check_deadline!(@data[:connect_timeout], :connect_read)` in each branch.

Part of the complexity is that we only want to use the request timeout if it is smaller than the operation's timeout. We don't want to use the request timeout just because one is defined, which is what we would get with `request_timeout || @data[:connect_timeout]`.

Reply:
This is what I think the select method would look like with your suggestion (please let me know if I misunderstood):

```ruby
    def select_with_timeout(socket, type)
      timeout_kind = type
      request_timeout = check_deadline! if @data.include?(:deadline)

      select = case type
      when :connect_read
        if request_timeout && request_timeout < @data[:connect_timeout]
          timeout_kind = type
        else
          request_timeout = nil
        end

        IO.select([socket], nil, nil, request_timeout || @data[:connect_timeout])
      when :connect_write
        # ...
        IO.select(nil, [socket], nil, @data[:connect_timeout])
      when :read
        # ...
        IO.select([socket], nil, nil, @data[:read_timeout])
      when :write
        # ...
        IO.select(nil, [socket], nil, @data[:write_timeout])
      end
      select || raise(Excon::Errors::Timeout.new("#{type} timeout reached"))
    end
```

Reply:
If it wasn't for the kind things would be a little bit simpler:

```ruby
    def select_with_timeout(socket, type)
      request_timeout = check_deadline! if @data.include?(:deadline)

      select = case type
      when :connect_read
        if !request_timeout.nil? && request_timeout >= @data[:connect_timeout]
          request_timeout = nil
        end

        IO.select([socket], nil, nil, request_timeout || @data[:connect_timeout])
      when :connect_write
        # ...
        IO.select(nil, [socket], nil, @data[:connect_timeout])
      when :read
        # ...
        IO.select([socket], nil, nil, @data[:read_timeout])
      when :write
        # ...
        IO.select(nil, [socket], nil, @data[:write_timeout])
      end
      select || raise(Excon::Errors::Timeout.new("#{type} timeout reached"))
    end
```

Reply:
At one point this is the solution I had. I didn't like it because we always paid a penalty of allocating those arrays and the method call regardless of whether a timeout was defined or not.

```ruby
    def select_with_timeout(socket, type)
      timeout_kind = type

      select = case type
      when :connect_read
        timeout_kind, timeout = check_deadline!(type, @data[:connect_timeout])
        IO.select([socket], nil, nil, timeout)
      when :connect_write
        timeout_kind, timeout = check_deadline!(type, @data[:connect_timeout])
        IO.select(nil, [socket], nil, timeout)
      when :read
        timeout_kind, timeout = check_deadline!(type, @data[:read_timeout])
        IO.select([socket], nil, nil, timeout)
      when :write
        timeout_kind, timeout = check_deadline!(type, @data[:write_timeout])
        IO.select(nil, [socket], nil, timeout)
      end

      select || raise(Excon::Errors::Timeout.new("#{timeout_kind} timeout reached"))
    end

    # Checks whether we have exceeded the request deadline.
    # Returns the remaining time in seconds as Numeric.
    def check_deadline!(type, timeout)
      return [type, timeout] unless @data.include?(:deadline)

      now = Process.clock_gettime(Process::CLOCK_MONOTONIC)
      deadline = @data[:deadline]
      
      raise raise(Excon::Errors::Timeout.new('request timeout reached')) if now >= deadline

      remaining  = deadline - now

      remaining < timeout ? [:request, remaining] : [type, timeout]
    end
```

Reply:
Those don't exactly line up with what I had in mind, but I think what I was thinking wouldn't work anyway now that I understand things a bit better from your explanation. That being said, I think it might be good to add some of that explanation as a comment to hopefully make it clearer. I'll add another comment in a second with what I mean.